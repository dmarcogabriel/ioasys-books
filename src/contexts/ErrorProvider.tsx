import React, {createContext} from 'react';
import {UNAUTHORIZED_STATUS} from '../constants';
import Toast from 'react-native-toast-message';

type ErrorData = {
  message: string;
};

export interface ApiError {
  errors: ErrorData;
}

interface ErrorContextValue {
  handleError: (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ) => Promise<void>;
}

export const ErrorContext = createContext<ErrorContextValue>({
  handleError: async () => {},
});

interface ErrorProviderProps {
  children: React.ReactNode;
}

export const ErrorProvider = ({children}: ErrorProviderProps): JSX.Element => {
  const handleError = async (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ): Promise<void> => {
    let errorMessage: string = '';
    if (status === UNAUTHORIZED_STATUS) {
      await logoutCallback();
      errorMessage = 'Sess√£o expirada';
    } else {
      Toast.show({
        position: 'bottom',
        text1: errorMessage || error.errors.message,
        type: 'error',
      });
    }
  };

  return (
    <ErrorContext.Provider value={{handleError}}>
      {children}
    </ErrorContext.Provider>
  );
};
