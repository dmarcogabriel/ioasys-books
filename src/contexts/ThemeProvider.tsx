import React from 'react';
import {ThemeProvider as SCThemeProvider} from 'styled-components/native';

interface ThemeProviderProps {
  children: React.ReactNode;
}

export const ThemeProvider = ({children}: ThemeProviderProps): JSX.Element => {
  const theme = {
    colors: {
      primary: '#B22E6F',
      dark: '#333333',
      light: '#ffffff',
      grey: '#999999',
    },
  };

  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};
