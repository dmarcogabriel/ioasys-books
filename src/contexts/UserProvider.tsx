import React, {createContext, useState, useEffect, useCallback} from 'react';
import {User} from '../@types/user.interface';
import {storeData, removeData, getData} from '../utils/storage';
import {AUTH_TOKEN_STORAGE_KEY} from '../constants';

interface UserContextValue {
  user: User | null;
  login: (user: User, token: string) => Promise<void>;
  logout: () => Promise<void>;
  token: string | null;
}

export const UserContext = createContext<UserContextValue>({
  user: null,
  token: null,
  login: async () => {},
  logout: async () => {},
});

interface UserProviderProps {
  children: React.ReactNode;
}

export const UserProvider = ({children}: UserProviderProps): JSX.Element => {
  const [user, setUser] = useState<User | null>(null);
  const [token, setToken] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  const login = async (data: User, authToken: string): Promise<void> => {
    setUser(data);
    setToken(authToken);
    await storeData(AUTH_TOKEN_STORAGE_KEY, authToken);
  };

  const logout = async () => {
    await removeData(AUTH_TOKEN_STORAGE_KEY);
    setUser(null);
    setToken(null);
  };

  const loadToken = useCallback(async () => {
    setIsLoading(true);
    try {
      const authToken = await getData<string>(AUTH_TOKEN_STORAGE_KEY);
      if (authToken) {
        setToken(authToken);
      }
    } catch (error) {
      setIsLoading(false);
    }
    setIsLoading(false);
  }, []);

  useEffect(() => {
    loadToken();
  }, [loadToken]);

  return (
    <UserContext.Provider value={{user, token, login, logout}}>
      {!isLoading && children}
    </UserContext.Provider>
  );
};
