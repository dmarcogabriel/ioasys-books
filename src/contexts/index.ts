export * from './UserProvider';
export * from './ThemeProvider';
export * from './ErrorProvider';
