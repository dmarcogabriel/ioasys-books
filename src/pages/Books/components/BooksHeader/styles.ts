import styled from 'styled-components/native';

export const BooksHeaderContainer = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  padding: 40px 16px 0;
  background: #e5e5e5;
`;

export const LogoutButton = styled.TouchableOpacity`
  border-radius: 100px;
  border: 1px solid #33333333;
  padding: 9px;
`;
