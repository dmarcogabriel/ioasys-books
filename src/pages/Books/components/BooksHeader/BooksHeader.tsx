import React from 'react';
import {BooksHeaderContainer, LogoutButton} from './styles';
import {HeaderTitle} from '../../../../components';
import MIcon from 'react-native-vector-icons/MaterialIcons';
import {useUser} from '../../../../hooks';

export const BooksHeader = () => {
  const {logout} = useUser();

  return (
    <BooksHeaderContainer>
      <HeaderTitle noPadding noMargin color="dark" />
      <LogoutButton testID="logoutButton" onPress={logout}>
        <MIcon name="logout" size={24} />
      </LogoutButton>
    </BooksHeaderContainer>
  );
};
