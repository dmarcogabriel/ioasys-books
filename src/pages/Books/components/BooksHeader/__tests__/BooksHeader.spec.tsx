import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {BooksHeader} from '../BooksHeader';
import {ThemeProvider} from '../../../../../contexts';
import renderer from 'react-test-renderer';

const mockLogout = jest.fn();
jest.mock('../../../../../hooks', () => ({
  useUser: () => ({
    logout: mockLogout,
  }),
}));

const ComponentWrapper = () => (
  <ThemeProvider>
    <BooksHeader />
  </ThemeProvider>
);

describe('Books/components/BooksHeader', () => {
  it('should pass on click logout', () => {
    const {getByTestId} = render(<ComponentWrapper />);
    const logoutButton = getByTestId('logoutButton');
    fireEvent.press(logoutButton);
    expect(mockLogout).toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    expect(renderer.create(<ComponentWrapper />).toJSON()).toMatchSnapshot();
  });
});
