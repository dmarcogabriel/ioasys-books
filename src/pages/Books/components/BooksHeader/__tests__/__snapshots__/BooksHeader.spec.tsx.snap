// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books/components/BooksHeader should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#e5e5e5",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 0,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 40,
      },
    ]
  }
>
  <View
    noMargin={true}
    noPadding={true}
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 0,
          "paddingTop": 0,
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/components/HeaderTitle/assets/logo-dark.png",
        }
      }
      style={
        Array [
          Object {},
        ]
      }
    />
    <Text
      color="dark"
      style={
        Array [
          Object {
            "color": "#333333",
            "fontSize": 28,
            "marginLeft": 16,
          },
        ]
      }
      testID="headerTitle"
    >
      Books
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderColor": "#33333333",
        "borderRadius": 100,
        "borderStyle": "solid",
        "borderWidth": 1,
        "opacity": 1,
        "paddingBottom": 9,
        "paddingLeft": 9,
        "paddingRight": 9,
        "paddingTop": 9,
      }
    }
    testID="logoutButton"
  >
    <Text>
      Icon
    </Text>
  </View>
</View>
`;
